engine/src/main/java/io/knotx/fragments/engine/TaskEngine.java

```diff
@@ -45,34 +45,41 @@
 
   private Single<TaskExecutionContext> processTask(TaskExecutionContext context) {
     traceEvent(context);
-    final Single<FragmentResult> result;
-    if (context.getCurrentNode().isComposite()) {
-      result = mapReduce(context);
-    } else {
-      result = execute(context);
+
+    if (!context.hasNext()) {
```

Comment:
I think you can combine that check in the next return statement - try to leverage single operators to emit different single